File: examples/lua/02_communication_and_messaging.lua
#!/usr/bin/env lua
--[[
QtForge Lua Bindings Example 2: Communication and Messaging

This example demonstrates inter-plugin communication using the QtForge
message bus system, including:
- Creating and configuring message buses
- Publishing and subscribing to messages
- Service contracts and discovery
- Request-response patterns
- Message priorities and delivery modes
--]]

-- Helper functions for output formatting
local function print_header(title)
    print("\n" .. string.rep("=", 50))
    print("📡 " .. title)
    print(string.rep("=", 50))
end

local function print_success(message)
    print("✅ " .. message)
end

local function print_warning(message)
    print("⚠️  " .. message)
end

local function print_error(message)
    print("❌ " .. message)
end

local function print_info(message)
    print("📋 " .. message)
end

local function print_status(message)
    print("📊 " .. message)
end

-- Check if QtForge communication bindings are available
local function check_communication_availability()
    if not qtforge then
        print_error("QtForge Lua bindings not available")
        return false
    end
    
    if not qtforge.communication then
        print_error("QtForge Communication module not available")
        return false
    end
    
    print_success("QtForge Communication bindings loaded successfully")
    return true
end

-- Demonstrate message bus creation and configuration
local function demonstrate_message_bus_creation()
    print_header("Creating Message Bus")
    
    if not qtforge.communication.create_message_bus then
        print_error("Message bus creation not available")
        return nil
    end
    
    local success, bus = pcall(qtforge.communication.create_message_bus)
    
    if not success then
        print_error("Failed to create message bus: " .. tostring(bus))
        return nil
    end
    
    print_success("Message bus created successfully")
    
    -- Check available methods
    local methods = {"publish", "subscribe", "unsubscribe"}
    local available_methods = {}
    
    for _, method in ipairs(methods) do
        if bus[method] then
            table.insert(available_methods, method)
        end
    end
    
    print_info("Available methods: " .. table.concat(available_methods, ", "))
    
    return bus
end

-- Demonstrate basic publish/subscribe messaging
local function demonstrate_basic_messaging(bus)
    print_header("Basic Publish/Subscribe Messaging")
    
    if not bus then
        print_error("No message bus available")
        return
    end
    
    -- Storage for received messages
    local received_messages = {}
    
    -- Callback function for received messages
    local function message_callback(message)
        print("📥 Received message: " .. tostring(message))
        table.insert(received_messages, message)
    end
    
    -- Subscribe to a topic
    local topic = "example.notifications"
    print("🔔 Subscribing to topic: " .. topic)
    
    if bus.subscribe then
        local sub_success, sub_result = pcall(function()
            return bus:subscribe(topic, message_callback)
        end)
        
        if sub_success then
            print_success("Successfully subscribed to topic")
        else
            print_warning("Subscription failed: " .. tostring(sub_result))
            return
        end
    else
        print_warning("Subscribe method not available")
        return
    end
    
    -- Create and publish messages
    if qtforge.communication.BasicMessage then
        print("\n📤 Publishing messages to topic: " .. topic)
        
        local messages = {
            "Hello, World!",
            "This is a test message",
            {type = "notification", data = "Important update"},
            42  -- Test with different data types
        }
        
        for i, msg_data in ipairs(messages) do
            local msg_success, message = pcall(function()
                return qtforge.communication.BasicMessage(topic, msg_data)
            end)
            
            if msg_success then
                local pub_success, pub_result = pcall(function()
                    return bus:publish(message)
                end)
                
                if pub_success then
                    print("  ✅ Published message " .. i .. ": " .. tostring(msg_data))
                else
                    print("  ❌ Failed to publish message " .. i .. ": " .. tostring(pub_result))
                end
            else
                print("  ❌ Failed to create message " .. i .. ": " .. tostring(message))
            end
            
            -- Small delay for processing
            os.execute("sleep 0.1")
        end
    else
        print_warning("BasicMessage class not available")
    end
    
    -- Give some time for message processing
    os.execute("sleep 0.5")
    
    print_status("Total messages received: " .. #received_messages)
    
    -- Unsubscribe
    if bus.unsubscribe then
        local unsub_success, unsub_result = pcall(function()
            return bus:unsubscribe(topic, message_callback)
        end)
        
        if unsub_success then
            print_success("Successfully unsubscribed from topic")
        else
            print_warning("Unsubscription failed: " .. tostring(unsub_result))
        end
    end
end

-- Demonstrate message properties and metadata
local function demonstrate_message_properties()
    print_header("Message Properties and Metadata")
    
    if not qtforge.communication.BasicMessage then
        print_error("BasicMessage class not available")
        return
    end
    
    local topic = "example.properties"
    local data = {user = "alice", action = "login", timestamp = os.time()}
    
    local success, message = pcall(function()
        return qtforge.communication.BasicMessage(topic, data)
    end)
    
    if not success then
        print_error("Failed to create message: " .. tostring(message))
        return
    end
    
    print_success("Created message with basic properties")
    
    -- Test property access
    if message.topic then
        print_info("Message topic: " .. tostring(message.topic))
    end
    
    if message.data then
        print_info("Message data: " .. tostring(message.data))
    end
    
    -- Test metadata operations
    if message.set_metadata then
        print("\n🏷️  Setting message metadata...")
        
        local metadata_items = {
            {"priority", "high"},
            {"sender", "user_service"},
            {"correlation_id", "12345"},
            {"timestamp", tostring(os.time())}
        }
        
        for _, item in ipairs(metadata_items) do
            local key, value = item[1], item[2]
            
            local meta_success, meta_result = pcall(function()
                return message:set_metadata(key, value)
            end)
            
            if meta_success then
                print("  ✅ Set " .. key .. ": " .. value)
            else
                print("  ❌ Failed to set " .. key .. ": " .. tostring(meta_result))
            end
        end
    end
    
    -- Test metadata retrieval
    if message.get_metadata then
        print("\n📋 Retrieving message metadata...")
        
        local metadata_keys = {"priority", "sender", "correlation_id", "timestamp"}
        
        for _, key in ipairs(metadata_keys) do
            local get_success, value = pcall(function()
                return message:get_metadata(key)
            end)
            
            if get_success then
                if value then
                    print("  ✅ " .. key .. ": " .. tostring(value))
                else
                    print("  ⚠️  " .. key .. ": not set")
                end
            else
                print("  ❌ Failed to get " .. key .. ": " .. tostring(value))
            end
        end
    end
    
    return message
end

-- Demonstrate message priorities and delivery modes
local function demonstrate_message_priorities()
    print_header("Message Priorities and Delivery Modes")
    
    -- Test MessagePriority enum
    if MessagePriority then
        print("📊 Available message priorities:")
        local priorities = {"Low", "Normal", "High", "Critical"}
        
        for _, priority in ipairs(priorities) do
            if MessagePriority[priority] then
                local value = MessagePriority[priority]
                print("  • " .. priority .. ": " .. tostring(value))
            end
        end
    end
    
    -- Test DeliveryMode enum
    if DeliveryMode then
        print("\n🚚 Available delivery modes:")
        local modes = {"Immediate", "Queued", "Persistent"}
        
        for _, mode in ipairs(modes) do
            if DeliveryMode[mode] then
                local value = DeliveryMode[mode]
                print("  • " .. mode .. ": " .. tostring(value))
            end
        end
    end
    
    -- Demonstrate setting priorities and delivery modes
    if qtforge.communication.BasicMessage then
        local success, message = pcall(function()
            return qtforge.communication.BasicMessage("priority.test", "High priority message")
        end)
        
        if success then
            -- Set priority if supported
            if message.set_priority and MessagePriority and MessagePriority.High then
                local priority_success, priority_result = pcall(function()
                    return message:set_priority(MessagePriority.High)
                end)
                
                if priority_success then
                    print_success("Set message priority to High")
                    
                    if message.get_priority then
                        local get_success, priority = pcall(function()
                            return message:get_priority()
                        end)
                        
                        if get_success then
                            print_status("Current priority: " .. tostring(priority))
                        end
                    end
                else
                    print_warning("Failed to set priority: " .. tostring(priority_result))
                end
            end
            
            -- Set delivery mode if supported
            if message.set_delivery_mode and DeliveryMode and DeliveryMode.Persistent then
                local mode_success, mode_result = pcall(function()
                    return message:set_delivery_mode(DeliveryMode.Persistent)
                end)
                
                if mode_success then
                    print_success("Set delivery mode to Persistent")
                    
                    if message.get_delivery_mode then
                        local get_success, mode = pcall(function()
                            return message:get_delivery_mode()
                        end)
                        
                        if get_success then
                            print_status("Current delivery mode: " .. tostring(mode))
                        end
                    end
                else
                    print_warning("Failed to set delivery mode: " .. tostring(mode_result))
                end
            end
        else
            print_error("Failed to create message for priority/delivery mode demo: " .. tostring(message))
        end
    end
end

-- Demonstrate service contracts and discovery
local function demonstrate_service_contracts()
    print_header("Service Contracts and Discovery")
    
    -- Test ServiceVersion
    if qtforge.communication.ServiceVersion then
        local version_success, version = pcall(function()
            return qtforge.communication.ServiceVersion(1, 2, 3)
        end)
        
        if version_success then
            print_success("Created service version 1.2.3")
            
            if version.major then
                print_status("Major version: " .. tostring(version.major))
            end
            if version.minor then
                print_status("Minor version: " .. tostring(version.minor))
            end
            if version.patch then
                print_status("Patch version: " .. tostring(version.patch))
            end
        else
            print_error("Failed to create service version: " .. tostring(version))
        end
    end
    
    -- Test ServiceMethodDescriptor
    if qtforge.communication.ServiceMethodDescriptor then
        local desc_success, descriptor = pcall(qtforge.communication.ServiceMethodDescriptor)
        
        if desc_success then
            print_success("Created service method descriptor")
            
            if descriptor.name then
                descriptor.name = "calculate"
                print_info("Method name: " .. tostring(descriptor.name))
            end
        else
            print_error("Failed to create method descriptor: " .. tostring(descriptor))
        end
    end
    
    -- Test ServiceContract
    if qtforge.communication.ServiceContract then
        local contract_success, contract = pcall(qtforge.communication.ServiceContract)
        
        if contract_success then
            print_success("Created service contract")
            
            -- Test adding methods if supported
            if contract.add_method and qtforge.communication.ServiceMethodDescriptor then
                local method_success, method = pcall(qtforge.communication.ServiceMethodDescriptor)
                
                if method_success and method.name then
                    method.name = "process_data"
                    
                    local add_success, add_result = pcall(function()
                        return contract:add_method(method)
                    end)
                    
                    if add_success then
                        print_success("Added method to service contract")
                    else
                        print_warning("Failed to add method: " .. tostring(add_result))
                    end
                end
            end
            
            -- Test contract validation
            if contract.validate then
                local validate_success, is_valid = pcall(function()
                    return contract:validate()
                end)
                
                if validate_success then
                    print_status("Contract validation result: " .. tostring(is_valid))
                else
                    print_warning("Contract validation failed: " .. tostring(is_valid))
                end
            end
        else
            print_error("Failed to create service contract: " .. tostring(contract))
        end
    end
    
    -- Test ServiceCapability enum
    if ServiceCapability then
        print("\n🛠️  Available service capabilities:")
        local capabilities = {"Synchronous", "Asynchronous", "Streaming", "Transactional"}
        
        for _, capability in ipairs(capabilities) do
            if ServiceCapability[capability] then
                local value = ServiceCapability[capability]
                print("  • " .. capability .. ": " .. tostring(value))
            end
        end
    end
end

-- Demonstrate request-response communication patterns
local function demonstrate_request_response()
    print_header("Request-Response Communication")
    
    -- Test Request creation
    if qtforge.communication.Request then
        local req_success, request = pcall(function()
            return qtforge.communication.Request("calculator_service", "add")
        end)
        
        if req_success then
            print_success("Created request for calculator_service.add")
            
            if request.service then
                print_info("Service: " .. tostring(request.service))
            end
            if request.method then
                print_info("Method: " .. tostring(request.method))
            end
        else
            print_error("Failed to create request: " .. tostring(request))
        end
    end
    
    -- Test Response creation
    if qtforge.communication.Response then
        local resp_success, response = pcall(qtforge.communication.Response)
        
        if resp_success then
            print_success("Created response object")
            
            -- Test setting result
            if response.set_result then
                local result_data = {sum = 42, status = "success"}
                
                local set_success, set_result = pcall(function()
                    return response:set_result(result_data)
                end)
                
                if set_success then
                    print_success("Set response result")
                    
                    if response.get_result then
                        local get_success, result = pcall(function()
                            return response:get_result()
                        end)
                        
                        if get_success then
                            print_status("Response result: " .. tostring(result))
                        end
                    end
                else
                    print_warning("Failed to set result: " .. tostring(set_result))
                end
            end
        else
            print_error("Failed to create response: " .. tostring(response))
        end
    end
    
    -- Demonstrate request-response with message bus
    local bus = nil
    if qtforge.communication.create_message_bus then
        local bus_success, bus_result = pcall(qtforge.communication.create_message_bus)
        if bus_success then
            bus = bus_result
        end
    end
    
    if bus and bus.send_request and qtforge.communication.Request then
        print("\n🔄 Testing request-response with message bus...")
        
        local req_success, request = pcall(function()
            return qtforge.communication.Request("echo_service", "echo")
        end)
        
        if req_success then
            local send_success, response = pcall(function()
                return bus:send_request(request, 1.0)  -- 1 second timeout
            end)
            
            if send_success then
                if response then
                    print_success("Received response from service")
                    if response.get_result then
                        local result_success, result = pcall(function()
                            return response:get_result()
                        end)
                        
                        if result_success then
                            print_status("Response data: " .. tostring(result))
                        end
                    end
                else
                    print_warning("No response received (service may not be available)")
                end
            else
                print_warning("Request-response failed: " .. tostring(response) .. " (this is expected if no service is running)")
            end
        end
    end
end

-- Demonstrate communication enumerations
local function demonstrate_communication_enums()
    print_header("Communication Enumerations")
    
    -- DeliveryMode enum
    if DeliveryMode then
        print("🚚 Delivery Modes:")
        local modes = {"Immediate", "Queued", "Persistent"}
        
        for _, mode in ipairs(modes) do
            if DeliveryMode[mode] then
                local value = DeliveryMode[mode]
                print("  • " .. mode .. ": " .. tostring(value))
            end
        end
    end
    
    -- MessagePriority enum
    if MessagePriority then
        print("\n⚡ Message Priorities:")
        local priorities = {"Low", "Normal", "High", "Critical"}
        
        for _, priority in ipairs(priorities) do
            if MessagePriority[priority] then
                local value = MessagePriority[priority]
                print("  • " .. priority .. ": " .. tostring(value))
            end
        end
    end
    
    -- ServiceCapability enum
    if ServiceCapability then
        print("\n🛠️  Service Capabilities:")
        local capabilities = {"Synchronous", "Asynchronous", "Streaming", "Transactional"}
        
        for _, capability in ipairs(capabilities) do
            if ServiceCapability[capability] then
                local value = ServiceCapability[capability]
                print("  • " .. capability .. ": " .. tostring(value))
            end
        end
    end
end

-- Demonstrate error handling in communication operations
local function demonstrate_error_handling()
    print_header("Communication Error Handling")
    
    -- Test various error conditions
    if qtforge.communication.create_message_bus then
        local success, bus = pcall(qtforge.communication.create_message_bus)
        
        if success then
            -- Test publishing None message
            print("🧪 Testing error handling with invalid parameters...")
            
            local pub_success, pub_result = pcall(function()
                return bus:publish(nil)
            end)
            
            if pub_success then
                print_warning("Unexpected success with nil message: " .. tostring(pub_result))
            else
                print_success("Correctly caught error for nil message: " .. tostring(pub_result))
            end
            
            -- Test subscribing with invalid topic
            local sub_success, sub_result = pcall(function()
                return bus:subscribe("", function() end)
            end)
            
            if sub_success then
                print_warning("Unexpected success with empty topic: " .. tostring(sub_result))
            else
                print_success("Correctly caught error for empty topic: " .. tostring(sub_result))
            end
        end
    end
    
    -- Test service not found error
    if qtforge.communication.create_message_bus and qtforge.communication.Request then
        local bus_success, bus = pcall(qtforge.communication.create_message_bus)
        
        if bus_success and bus.send_request then
            print("\n🧪 Testing service not found error...")
            
            local req_success, request = pcall(function()
                return qtforge.communication.Request("non_existent_service", "test_method")
            end)
            
            if req_success then
                local send_success, response = pcall(function()
                    return bus:send_request(request, 0.1)  -- Very short timeout
                end)
                
                if send_success then
                    if response and response.is_error then
                        local error_success, is_error = pcall(function()
                            return response:is_error()
                        end)
                        
                        if error_success and is_error then
                            print_success("Correctly received error response")
                        end
                    else
                        print_warning("No response for non-existent service (expected)")
                    end
                else
                    print_success("Correctly caught exception for non-existent service: " .. tostring(response))
                end
            end
        end
    end
end

-- Main demonstration function
local function main()
    print("QtForge Lua Bindings - Communication and Messaging Example")
    print(string.rep("=", 65))
    
    -- Check if communication bindings are available
    if not check_communication_availability() then
        return 1
    end
    
    -- Demonstrate each aspect of communication
    local bus = demonstrate_message_bus_creation()
    demonstrate_basic_messaging(bus)
    demonstrate_message_properties()
    demonstrate_message_priorities()
    demonstrate_service_contracts()
    demonstrate_request_response()
    demonstrate_communication_enums()
    demonstrate_error_handling()
    
    print("\n" .. string.rep("=", 65))
    print("🎉 Communication and Messaging Example Complete!")
    print(string.rep("=", 65))
    
    print("\n📚 Key Takeaways:")
    print("• Message buses enable decoupled inter-plugin communication")
    print("• Service contracts define clear interfaces between components")
    print("• Request-response patterns support synchronous interactions")
    print("• Message priorities and delivery modes control message handling")
    print("• Proper error handling ensures robust communication")
    
    print("\n🔗 Next Steps:")
    print("• Implement custom message handlers and filters")
    print("• Create service discovery and registration mechanisms")
    print("• Add message persistence and reliability features")
    print("• Explore advanced routing and transformation patterns")
    
    return 0
end

-- Run the main function
return main()

File: examples/python/02_communication_and_messaging.py
#!/usr/bin/env python3
"""
QtForge Python Bindings Example 2: Communication and Messaging

This example demonstrates inter-plugin communication using the QtForge
message bus system, including:
- Creating and configuring message buses
- Publishing and subscribing to messages
- Service contracts and discovery
- Request-response patterns
- Message priorities and delivery modes
"""

import sys
import time
import threading
from pathlib import Path

# Add the build directory to Python path
sys.path.insert(0, str(Path(__file__).parent.parent.parent / "build"))

try:
    import qtforge
    import qtforge.communication as comm
    print("✅ QtForge Communication bindings loaded successfully")
except ImportError as e:
    print(f"❌ Failed to import QtForge Communication: {e}")
    print("Make sure QtForge is built with Python bindings enabled")
    sys.exit(1)


def demonstrate_message_bus_creation():
    """Demonstrate creating and configuring a message bus."""
    print("\n" + "="*50)
    print("📡 Creating Message Bus")
    print("="*50)
    
    try:
        # Create a message bus
        bus = comm.create_message_bus()
        print("✅ Message bus created successfully")
        
        # Check available methods
        methods = ['publish', 'subscribe', 'unsubscribe']
        available_methods = []
        
        for method in methods:
            if hasattr(bus, method):
                available_methods.append(method)
        
        print(f"📋 Available methods: {', '.join(available_methods)}")
        
        return bus
        
    except Exception as e:
        print(f"❌ Failed to create message bus: {e}")
        return None


def demonstrate_basic_messaging(bus):
    """Demonstrate basic publish/subscribe messaging."""
    print("\n" + "="*50)
    print("📨 Basic Publish/Subscribe Messaging")
    print("="*50)
    
    if not bus:
        print("❌ No message bus available")
        return
    
    # Storage for received messages
    received_messages = []
    
    def message_callback(message):
        """Callback function for received messages."""
        print(f"📥 Received message: {message}")
        received_messages.append(message)
    
    try:
        # Subscribe to a topic
        topic = "example.notifications"
        print(f"🔔 Subscribing to topic: {topic}")
        
        if hasattr(bus, 'subscribe'):
            try:
                bus.subscribe(topic, message_callback)
                print("✅ Successfully subscribed to topic")
            except Exception as e:
                print(f"⚠️  Subscription failed: {e}")
                return
        
        # Create and publish messages
        if hasattr(comm, 'BasicMessage'):
            print(f"\n📤 Publishing messages to topic: {topic}")
            
            messages = [
                "Hello, World!",
                "This is a test message",
                {"type": "notification", "data": "Important update"},
                42  # Test with different data types
            ]
            
            for i, msg_data in enumerate(messages, 1):
                try:
                    message = comm.BasicMessage(topic, msg_data)
                    bus.publish(message)
                    print(f"  ✅ Published message {i}: {msg_data}")
                    time.sleep(0.1)  # Small delay for processing
                except Exception as e:
                    print(f"  ❌ Failed to publish message {i}: {e}")
        
        # Give some time for message processing
        time.sleep(0.5)
        
        print(f"\n📊 Total messages received: {len(received_messages)}")
        
        # Unsubscribe
        if hasattr(bus, 'unsubscribe'):
            try:
                bus.unsubscribe(topic, message_callback)
                print("✅ Successfully unsubscribed from topic")
            except Exception as e:
                print(f"⚠️  Unsubscription failed: {e}")
        
    except Exception as e:
        print(f"❌ Basic messaging demonstration failed: {e}")


def demonstrate_message_properties():
    """Demonstrate message properties and metadata."""
    print("\n" + "="*50)
    print("🏷️  Message Properties and Metadata")
    print("="*50)
    
    if not hasattr(comm, 'BasicMessage'):
        print("❌ BasicMessage class not available")
        return
    
    try:
        # Create a message with basic properties
        topic = "example.properties"
        data = {"user": "alice", "action": "login", "timestamp": time.time()}
        
        message = comm.BasicMessage(topic, data)
        print("✅ Created message with basic properties")
        
        # Test property access
        if hasattr(message, 'topic'):
            print(f"📋 Message topic: {message.topic}")
        
        if hasattr(message, 'data'):
            print(f"📋 Message data: {message.data}")
        
        # Test metadata operations
        if hasattr(message, 'set_metadata'):
            print("\n🏷️  Setting message metadata...")
            
            metadata_items = [
                ("priority", "high"),
                ("sender", "user_service"),
                ("correlation_id", "12345"),
                ("timestamp", str(int(time.time())))
            ]
            
            for key, value in metadata_items:
                try:
                    message.set_metadata(key, value)
                    print(f"  ✅ Set {key}: {value}")
                except Exception as e:
                    print(f"  ❌ Failed to set {key}: {e}")
        
        # Test metadata retrieval
        if hasattr(message, 'get_metadata'):
            print("\n📋 Retrieving message metadata...")
            
            for key, expected_value in metadata_items:
                try:
                    value = message.get_metadata(key)
                    if value == expected_value:
                        print(f"  ✅ {key}: {value}")
                    else:
                        print(f"  ⚠️  {key}: {value} (expected: {expected_value})")
                except Exception as e:
                    print(f"  ❌ Failed to get {key}: {e}")
        
        return message
        
    except Exception as e:
        print(f"❌ Message properties demonstration failed: {e}")
        return None


def demonstrate_message_priorities():
    """Demonstrate message priorities and delivery modes."""
    print("\n" + "="*50)
    print("⚡ Message Priorities and Delivery Modes")
    print("="*50)
    
    # Test MessagePriority enum
    if hasattr(comm, 'MessagePriority'):
        print("📊 Available message priorities:")
        priorities = ['Low', 'Normal', 'High', 'Critical']
        
        for priority in priorities:
            if hasattr(comm.MessagePriority, priority):
                value = getattr(comm.MessagePriority, priority)
                print(f"  • {priority}: {value}")
    
    # Test DeliveryMode enum
    if hasattr(comm, 'DeliveryMode'):
        print("\n🚚 Available delivery modes:")
        modes = ['Immediate', 'Queued', 'Persistent']
        
        for mode in modes:
            if hasattr(comm.DeliveryMode, mode):
                value = getattr(comm.DeliveryMode, mode)
                print(f"  • {mode}: {value}")
    
    # Demonstrate setting priorities and delivery modes
    if hasattr(comm, 'BasicMessage'):
        try:
            message = comm.BasicMessage("priority.test", "High priority message")
            
            # Set priority if supported
            if (hasattr(message, 'set_priority') and 
                hasattr(comm, 'MessagePriority') and 
                hasattr(comm.MessagePriority, 'High')):
                
                message.set_priority(comm.MessagePriority.High)
                print("✅ Set message priority to High")
                
                if hasattr(message, 'get_priority'):
                    priority = message.get_priority()
                    print(f"📊 Current priority: {priority}")
            
            # Set delivery mode if supported
            if (hasattr(message, 'set_delivery_mode') and 
                hasattr(comm, 'DeliveryMode') and 
                hasattr(comm.DeliveryMode, 'Persistent')):
                
                message.set_delivery_mode(comm.DeliveryMode.Persistent)
                print("✅ Set delivery mode to Persistent")
                
                if hasattr(message, 'get_delivery_mode'):
                    mode = message.get_delivery_mode()
                    print(f"📊 Current delivery mode: {mode}")
            
        except Exception as e:
            print(f"❌ Priority/delivery mode demonstration failed: {e}")


def demonstrate_service_contracts():
    """Demonstrate service contracts and discovery."""
    print("\n" + "="*50)
    print("📋 Service Contracts and Discovery")
    print("="*50)
    
    # Test ServiceVersion
    if hasattr(comm, 'ServiceVersion'):
        try:
            version = comm.ServiceVersion(1, 2, 3)
            print("✅ Created service version 1.2.3")
            
            if hasattr(version, 'major'):
                print(f"📊 Major version: {version.major}")
            if hasattr(version, 'minor'):
                print(f"📊 Minor version: {version.minor}")
            if hasattr(version, 'patch'):
                print(f"📊 Patch version: {version.patch}")
                
        except Exception as e:
            print(f"❌ Failed to create service version: {e}")
    
    # Test ServiceMethodDescriptor
    if hasattr(comm, 'ServiceMethodDescriptor'):
        try:
            descriptor = comm.ServiceMethodDescriptor()
            print("✅ Created service method descriptor")
            
            if hasattr(descriptor, 'name'):
                descriptor.name = "calculate"
                print(f"📋 Method name: {descriptor.name}")
                
        except Exception as e:
            print(f"❌ Failed to create method descriptor: {e}")
    
    # Test ServiceContract
    if hasattr(comm, 'ServiceContract'):
        try:
            contract = comm.ServiceContract()
            print("✅ Created service contract")
            
            # Test adding methods if supported
            if (hasattr(contract, 'add_method') and 
                hasattr(comm, 'ServiceMethodDescriptor')):
                
                method = comm.ServiceMethodDescriptor()
                if hasattr(method, 'name'):
                    method.name = "process_data"
                
                try:
                    contract.add_method(method)
                    print("✅ Added method to service contract")
                except Exception as e:
                    print(f"⚠️  Failed to add method: {e}")
            
            # Test contract validation
            if hasattr(contract, 'validate'):
                try:
                    is_valid = contract.validate()
                    print(f"📊 Contract validation result: {is_valid}")
                except Exception as e:
                    print(f"⚠️  Contract validation failed: {e}")
                    
        except Exception as e:
            print(f"❌ Failed to create service contract: {e}")
    
    # Test ServiceCapability enum
    if hasattr(comm, 'ServiceCapability'):
        print("\n🛠️  Available service capabilities:")
        capabilities = ['Synchronous', 'Asynchronous', 'Streaming', 'Transactional']
        
        for capability in capabilities:
            if hasattr(comm.ServiceCapability, capability):
                value = getattr(comm.ServiceCapability, capability)
                print(f"  • {capability}: {value}")


def demonstrate_request_response():
    """Demonstrate request-response communication patterns."""
    print("\n" + "="*50)
    print("🔄 Request-Response Communication")
    print("="*50)
    
    # Test Request creation
    if hasattr(comm, 'Request'):
        try:
            request = comm.Request("calculator_service", "add")
            print("✅ Created request for calculator_service.add")
            
            if hasattr(request, 'service'):
                print(f"📋 Service: {request.service}")
            if hasattr(request, 'method'):
                print(f"📋 Method: {request.method}")
                
        except Exception as e:
            print(f"❌ Failed to create request: {e}")
    
    # Test Response creation
    if hasattr(comm, 'Response'):
        try:
            response = comm.Response()
            print("✅ Created response object")
            
            # Test setting result
            if hasattr(response, 'set_result'):
                response.set_result({"sum": 42, "status": "success"})
                print("✅ Set response result")
                
                if hasattr(response, 'get_result'):
                    result = response.get_result()
                    print(f"📊 Response result: {result}")
                    
        except Exception as e:
            print(f"❌ Failed to create response: {e}")
    
    # Demonstrate request-response with message bus
    bus = comm.create_message_bus() if hasattr(comm, 'create_message_bus') else None
    
    if bus and hasattr(bus, 'send_request') and hasattr(comm, 'Request'):
        print("\n🔄 Testing request-response with message bus...")
        
        try:
            request = comm.Request("echo_service", "echo")
            
            # Test with timeout
            response = bus.send_request(request, timeout=1.0)
            
            if response:
                print("✅ Received response from service")
                if hasattr(response, 'get_result'):
                    result = response.get_result()
                    print(f"📊 Response data: {result}")
            else:
                print("⚠️  No response received (service may not be available)")
                
        except Exception as e:
            print(f"⚠️  Request-response failed: {e} (this is expected if no service is running)")


def demonstrate_async_communication():
    """Demonstrate asynchronous communication patterns."""
    print("\n" + "="*50)
    print("🔀 Asynchronous Communication")
    print("="*50)
    
    bus = comm.create_message_bus() if hasattr(comm, 'create_message_bus') else None
    
    if not bus:
        print("❌ No message bus available for async demo")
        return
    
    # Test async request-response
    if hasattr(bus, 'send_request_async') and hasattr(comm, 'Request'):
        print("🔄 Testing asynchronous request-response...")
        
        try:
            request = comm.Request("async_service", "process")
            future = bus.send_request_async(request)
            
            if future:
                print("✅ Async request sent, future received")
                
                # Test waiting for result
                if hasattr(future, 'wait'):
                    try:
                        result = future.wait(timeout=1.0)
                        print(f"📊 Async result: {result}")
                    except Exception as e:
                        print(f"⚠️  Async wait failed: {e}")
                        
        except Exception as e:
            print(f"⚠️  Async request failed: {e}")
    
    # Demonstrate threaded message handling
    print("\n🧵 Testing threaded message handling...")
    
    messages_received = []
    
    def threaded_subscriber(topic_suffix):
        """Function to run in separate thread for message handling."""
        topic = f"async.test.{topic_suffix}"
        
        def callback(message):
            messages_received.append(f"Thread-{topic_suffix}: {message}")
        
        try:
            if hasattr(bus, 'subscribe'):
                bus.subscribe(topic, callback)
                time.sleep(0.5)  # Wait for messages
        except Exception as e:
            print(f"⚠️  Thread {topic_suffix} subscription failed: {e}")
    
    # Start multiple subscriber threads
    threads = []
    for i in range(3):
        thread = threading.Thread(target=threaded_subscriber, args=(i,))
        threads.append(thread)
        thread.start()
    
    # Give threads time to subscribe
    time.sleep(0.2)
    
    # Publish messages to different topics
    if hasattr(comm, 'BasicMessage'):
        for i in range(3):
            topic = f"async.test.{i}"
            try:
                message = comm.BasicMessage(topic, f"Message for thread {i}")
                bus.publish(message)
                print(f"📤 Published to {topic}")
            except Exception as e:
                print(f"❌ Failed to publish to {topic}: {e}")
    
    # Wait for threads to complete
    for thread in threads:
        thread.join(timeout=2.0)
    
    print(f"📊 Total messages received by threads: {len(messages_received)}")
    for msg in messages_received:
        print(f"  📥 {msg}")


def main():
    """Main demonstration function."""
    print("QtForge Python Bindings - Communication and Messaging Example")
    print("=" * 65)
    
    # Demonstrate each aspect of communication
    bus = demonstrate_message_bus_creation()
    demonstrate_basic_messaging(bus)
    demonstrate_message_properties()
    demonstrate_message_priorities()
    demonstrate_service_contracts()
    demonstrate_request_response()
    demonstrate_async_communication()
    
    print("\n" + "="*65)
    print("🎉 Communication and Messaging Example Complete!")
    print("="*65)
    
    print("\n📚 Key Takeaways:")
    print("• Message buses enable decoupled inter-plugin communication")
    print("• Service contracts define clear interfaces between components")
    print("• Request-response patterns support synchronous interactions")
    print("• Message priorities and delivery modes control message handling")
    print("• Asynchronous patterns improve application responsiveness")
    
    print("\n🔗 Next Steps:")
    print("• Implement custom message handlers and filters")
    print("• Create service discovery and registration mechanisms")
    print("• Add message persistence and reliability features")
    print("• Explore advanced routing and transformation patterns")


if __name__ == "__main__":
    main()
